File: back/src/test/java/com/tdevilleduc/urthehero/back/service/StoryServiceTest.java
package com.tdevilleduc.urthehero.back.service;

import com.tdevilleduc.urthehero.back.AbstractTest;
import com.tdevilleduc.urthehero.back.exceptions.StoryNotFoundException;
import com.tdevilleduc.urthehero.back.model.Story;
import com.tdevilleduc.urthehero.back.service.impl.StoryService;
import io.github.resilience4j.circuitbreaker.CircuitBreaker.State;
import io.github.resilience4j.circuitbreaker.CircuitBreakerRegistry;
import io.vavr.collection.Stream;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.web.context.WebApplicationContext;

import java.util.List;
import java.util.Optional;

import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.notNullValue;
import static org.junit.jupiter.api.Assertions.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static org.springframework.test.web.servlet.setup.MockMvcBuilders.webAppContextSetup;

@Slf4j
@ExtendWith(SpringExtension.class)
@SpringBootTest
class StoryServiceTest extends AbstractTest {

    @Autowired
    private StoryService storyService;

    @Test
    void exists_thenCorrect() {
        Integer storyId = 1;
        boolean exists = storyService.exists(storyId);
        Assertions.assertTrue(exists);
    }

    @Test
    void notExists_thenCorrect() {
        Integer storyId = 41;
        boolean notExists = storyService.notExists(storyId);
        Assertions.assertTrue(notExists);
    }

    @Test
    void findByPageId_thenCorrect() {
        Integer storyId = 1;
        Optional<Story> optional = storyService.findById(storyId);

        Assertions.assertTrue(optional.isPresent());
        Story story = optional.get();
        Assertions.assertEquals(Integer.valueOf(1), story.getStoryId());
        Assertions.assertEquals("Ulysse", story.getTitle());
        Assertions.assertEquals(Integer.valueOf(1), story.getAuthorId());
        Assertions.assertEquals(Integer.valueOf(1), story.getFirstPageId());
        Assertions.assertEquals(Long.valueOf(3), story.getNumberOfReaders());
    }

    @Test
    void findById_withIdNull() {
        // with resilience4j, IllegalArgumentException is catched and fallback method 'emptyStory' is called
//        Assertions.assertThrows(IllegalArgumentException.class, () -> storyService.findById(null));
        assertEquals(Optional.empty(), storyService.findById(null));
    }
}

File: back/src/test/java/com/tdevilleduc/urthehero/back/service/StoryServiceTest.kt
package com.tdevilleduc.urthehero.back.service

import com.tdevilleduc.urthehero.back.dao.StoryDao
import com.tdevilleduc.urthehero.back.exceptions.StoryNotFoundException
import com.tdevilleduc.urthehero.back.service.impl.PageService
import com.tdevilleduc.urthehero.back.service.impl.ProgressionService
import com.tdevilleduc.urthehero.back.service.impl.StoryService
import com.tdevilleduc.urthehero.back.util.TestUtil
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import java.util.*

@ExtendWith(MockitoExtension::class)
internal class StoryServiceTest {
    private val random: Random = Random()

    @InjectMocks
    private lateinit var storyService: StoryService
    @Mock
    private lateinit var storyDao: StoryDao
    @Mock
    private lateinit var progressionService: ProgressionService
    @Mock
    private lateinit var pageService: PageService

    @Test
    fun test_exists_withIdNull() {
        val storyId = null
        val exists = storyService.exists(storyId)
        Assertions.assertFalse(exists)
    }

    @Test
    fun test_notExists_thenSuccess() {
        val storyId = random.nextInt()
        Mockito.`when`(storyDao.findById(storyId)).thenReturn(Optional.empty())

        val notExists = storyService.notExists(storyId)
        Assertions.assertTrue(notExists)
    }

    @Test
    fun test_findByPageId_thenCorrect() {
        val expectedPage = TestUtil.createStory()
        Mockito.`when`(storyDao.findById(expectedPage.storyId)).thenReturn(Optional.of(expectedPage))

        val story = storyService.findById(expectedPage.storyId)
        Assertions.assertNotNull(story)
        Assertions.assertEquals(expectedPage.storyId, story.storyId)
        Assertions.assertEquals(expectedPage.title, story.title)
        Assertions.assertEquals(expectedPage.authorId, story.authorId)
        Assertions.assertEquals(expectedPage.firstPageId, story.firstPageId)
        Assertions.assertEquals(expectedPage.numberOfReaders, story.numberOfReaders)
        Assertions.assertEquals(expectedPage.numberOfPages, story.numberOfPages)
    }

    @Test
    fun test_findById_thenNotFound() {
        val storyId = random.nextInt()
        Mockito.`when`(storyDao.findById(storyId)).thenReturn(Optional.empty())

        Assertions.assertThrows(StoryNotFoundException::class.java) { storyService.findById(storyId) }
    }
}